In [5]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

In [6]:
%%sql
SELECT * FROM sqlite_master WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

In [7]:
%%sql
SELECT *
FROM facts
LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Here are the descriptions for some of the columns:

* name - The name of the country.
* area - The total land and sea area of the country.
* population - The country's population.
* population_growth- The country's population growth as a percentage.
* birth_rate - The country's birth rate, or the number of births a year per 1,000 people.
* death_rate - The country's death rate, or the number of death a year per 1,000 people.
* area- The country's total area (both land and water).
* area_land - The country's land area in square kilometers.
* area_water - The country's waterarea in square kilometers.

Let's start by calculating some summary statistics and look for any outlier countries.

In [13]:
#calculate some summary statistics and look for any outlier countries.

In [14]:
%%sql
SELECT MIN(population), 
        MAX(population), 
        MIN(population_growth),
        MAX(population_growth)
FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

It's easy to see from the last statement that there is some data that needs to be cleaned.

* there's a country with a population of 0
* there's a country with a population of 7256490011 (close to the entire world population!)

Lets us zoom in on just these countries.

In [15]:
%%sql
SELECT *
FROM facts
WHERE population = (SELECT MIN(population)
                    FROM facts)
;

 * sqlite:///factbook.db
Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

In [16]:
%%sql
SELECT *
FROM facts
WHERE population = (SELECT MAX(population)
                    FROM facts)
;

 * sqlite:///factbook.db
Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

Looks like the dataset contains a row for the whole world, which explains the population of over 7.2 billion. It also seems like the table contains a row for Antarctica, which explains the population of 0.

Let's calculate some averages:

In [17]:
%%sql
SELECT  CAST(AVG(population) AS INT) AS "Average Population",
        ROUND(AVG(area),2) AS "Average Area"
FROM    facts
WHERE   population < (SELECT MAX(population)
                    FROM facts)
        AND population > (SELECT MIN(population)
                    FROM facts)
;

 * sqlite:///factbook.db
Done.


[(32377011, 582949.85)]

Now we can find countries that are densely populated. We'll identify countries that have:

- Above average values for population.
- Below average values for area.

In [18]:
%%sql
SELECT *
FROM facts 
WHERE population > (SELECT  CAST(AVG(population) AS INT) AS "Average Population"
                    FROM    facts
                    WHERE   population < (SELECT MAX(population)
                                        FROM facts)
                            AND population > (SELECT MIN(population)
                                        FROM facts)
                   )
      AND
      area < (SELECT  ROUND(AVG(area),2) AS "Average Area"
              FROM    facts
              WHERE   population < (SELECT MAX(population)
                                    FROM facts)
                        AND population > (SELECT MIN(population)
                        FROM facts)
             )
;

 * sqlite:///factbook.db
Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (88, 'ke', 'Kenya', 580367, 569140, 11227, 45925301, 1.93, 26.4, 6.89, 0.22),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

Here are some next steps for to explore:

- Which countries will add the most people to their population next year?
- Which countries have a higher death rate than birth rate?

Which countries have the highest ratios of water to land?

In [29]:
%%sql
SELECT *
FROM facts
WHERE (area_land > 0) AND (area_water > 0) AND (population > 0)
ORDER BY (CAST(area_water AS FLOAT) / CAST(area_land AS FLOAT)) DESC

 * sqlite:///factbook.db
Done.


[(247, 'vq', 'Virgin Islands', 1910, 346, 1564, 103574, 0.59, 10.31, 8.54, 7.67),
 (246, 'rq', 'Puerto Rico', 13791, 8870, 4921, 3598357, 0.6, 10.86, 8.67, 8.15),
 (12, 'bf', 'Bahamas, The', 13880, 10010, 3870, 324597, 0.85, 15.5, 7.05, 0.0),
 (71, 'pu', 'Guinea-Bissau', 36125, 28120, 8005, 1726170, 1.91, 33.38, 14.33, 0.0),
 (106, 'mi', 'Malawi', 118484, 94080, 24404, 17964697, 3.32, 41.56, 8.41, 0.0),
 (125, 'nl', 'Netherlands', 41543, 33893, 7650, 16947904, 0.41, 10.83, 8.66, 1.95),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (56, 'er', 'Eritrea', 117600, 101000, 16600, 6527689, 2.25, 30.0, 7.52, 0.0),
 (99, 'li', 'Liberia', 111369, 96320, 15049, 4195666, 2.47, 34.41, 9.69, 0.0),
 (14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (63, 'ga', 'Gambia, The', 11300, 10120, 1180, 1967709, 2.16, 30.86, 7.15, 2.12),
 (196, 'tw', 'Taiwan', 35980, 32260, 3720, 23415126, 0.23, 8.47, 7.11, 0.89),
 (60, 'fi', 'Finland', 338145, 303815, 34330, 5476922, 0.4, 10.72, 9.83, 3.1),
 (77, 'in', 'India', 3287263, 2973193, 314070, 1251695584, 1.22, 19.55, 7.32, 0.04),
 (32, 'ca', 'Canada', 9984670, 9093507, 891163, 35099836, 0.75, 10.28, 8.42, 5.66),
 (168, 'sw', 'Sweden', 450295, 410335, 39960, 9801616, 0.8, 11.99, 9.4, 5.42),
 (38, 'co', 'Colombia', 1138910, 1038700, 100210, 46736728, 1.04, 16.47, 5.4, 0.64),
 (25, 'bx', 'Brunei', 5765, 5265, 500, 429646, 1.62, 17.32, 3.52, 2.43),
 (72, 'gy', 'Guyana', 214969, 196849, 18120, 735222, 0.02, 15.59, 7.32, 8.06),
 (209, 'fp', 'French Polynesia', 4167, 3827, 340, 282703, 0.94, 15.22, 5.02, 0.84),
 (127, 'nu', 'Nicaragua', 130370, 119990, 10380, 5907881, 1.0, 18.03, 5.08, 3.0),
 (29, 'by', 'Burundi', 27830, 25680, 2150, 10742276, 3.28, 42.01, 9.27, 0.0),
 (79, 'ir', 'Iran', 1648195, 1531595, 116600, 81824270, 1.2, 17.99, 5.94, 0.07),
 (186, 'us', 'United States', 9826675, 9161966, 664709, 321368864, 0.78, 12.49, 8.15, 3.86),
 (172, 'tz', 'Tanzania', 947300, 885800, 61500, 51045882, 2.79, 36.39, 8.0, 0.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3),
 (144, 'rw', 'Rwanda', 26338, 24668, 1670, 12661733, 2.56, 33.75, 8.96, 0.85),
 (57, 'en', 'Estonia', 45228, 42388, 2840, 1265420, 0.55, 10.51, 12.4, 3.6),
 (130, 'no', 'Norway', 323802, 304282, 19520, 5207689, 1.13, 12.14, 8.12, 7.25),
 (8, 'am', 'Armenia', 29743, 28203, 1540, 3056382, 0.15, 13.61, 9.34, 5.8),
 (178, 'ts', 'Tunisia', 163610, 155360, 8250, 11037225, 0.89, 16.64, 5.98, 1.73),
 (188, 'uz', 'Uzbekistan', 447400, 425400, 22000, 29199942, 0.93, 17.0, 5.3, 2.37),
 (78, 'id', 'Indonesia', 1904569, 1811569, 93000, 255993674, 0.92, 16.72, 6.37, 1.16),
 (166, 'ns', 'Suriname', 163820, 156000, 7820, 579633, 1.08, 16.34, 6.13, 0.56),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (66, 'gh', 'Ghana', 238533, 227533, 11000, 26327649, 2.18, 31.09, 7.22, 2.02),
 (11, 'aj', 'Azerbaijan', 86600, 82629, 3971, 9780780, 0.96, 16.64, 7.07, 0.0),
 (175, 'to', 'Togo', 56785, 54385, 2400, 7552318, 2.69, 34.13, 7.26, 0.0),
 (143, 'rs', 'Russia', 17098242, 16377742, 720500, 142423773, 0.04, 11.6, 13.69, 1.69),
 (94, 'kg', 'Kyrgyzstan', 199951, 191801, 8150, 5664939, 1.11, 22.98, 6.65, 5.22),
 (176, 'tn', 'Tonga', 747, 717, 30, 106501, 0.03, 23.0, 4.85, 17.84),
 (183, 'up', 'Ukraine', 603550, 579330, 24220, 44429471, 0.6, 10.72, 14.46, 2.25),
 (102, 'lh', 'Lithuania', 65300, 62680, 2620, 2884433, 1.04, 10.1, 14.27, 6.27),
 (254, 'we', 'West Bank', 5860, 5640, 220, 2785366, 1.95, 22.99, 3.5, 0.0),
 (62, 'gb', 'Gabon', 267667, 257667, 10000, 1705336, 1.93, 34.49, 13.12, 2.03),
 (180, 'tx', 'Turkmenistan', 488100, 469930, 18170, 5231422, 1.14, 19.4, 6.13, 1.84),
 (75, 'hu', 'Hungary', 93028, 89608, 3420, 9897541, 0.22, 9.16, 12.73, 1.33),
 (96, 'lg', 'Latvia', 64589, 62249, 2340, 1986705, 1.06, 10.0, 14.31, 6.26),
 (142, 'ro', 'Romania', 238391, 229891, 8500, 21666350, 0.3, 9.14, 11.9, 0.24),
 (85, 'ja', 'Japan', 377915, 364485, 

Which countrie(s) have more water than land?

In [32]:
%%sql
SELECT name, area_water, area_land, population
FROM facts
WHERE (area_land > 0) AND (area_water > 0) AND (population > 0) AND (area_water > area_land)
;

 * sqlite:///factbook.db
Done.


[('Virgin Islands', 1564, 346, 103574)]